File: Easy/696_Count_binary_Substrings.cpp
///first attempt 5.78%faster, 100%less memory
//Idea:find '10' and '01' then expand them forward and backward with 1 and 0 respectively
class Solution {
public:
    int countBinarySubstrings(string s) {
    int k=-1;
    int k1=0;
    int count=0;
    while(k1!=-1)
    {
        k1=s.find("10",k+1);//find all "10"
        k=k1;
        if(k1==-1)
        break;
        count++;
        int i=1,j=2;
       
        while(k1-i>=0&&k1+j<s.size())//check wether backwards are 1s and forward are 0s
        {
           
            if(s[k1-i]=='1'&&s[k1+j]=='0')
            {
                
                 count++;
                 i++;
                 j++;
            }
           
            else
            break;
        }
    }
    
    k=-1;
    k1=0;
   while(k1!=-1)
    {
        k1=s.find("01",k+1);//find all "01" and rest similar as above
        k=k1;
        if(k1==-1)
        break;
        count++;
        int i=1,j=2;
        while(k1-i>=0&&k1+j<s.size())
        {
            if(s[k1-i]=='0'&&s[k1+j]=='1')
            {
                 count++;
                 i++;
                 j++;
            }
           
            else
            break;
        }
    }
    
    return count;
    }
};

File: Easy_python/696_Count_binary_substrings.py
class Solution:
#找出01和10字符串，然后向左右拓展0和1
    def countBinarySubstrings(self, s: str) -> int:
        k=0
        count=0
        while s.find('01',k)!=-1:
            index=s.find('01',k)
            count=count+1
            i=index-1
            j=index+2
            while j<len(s) and s[i]=='0' and s[j]=='1':
                count=count+1
                i=i-1
                j=j+1
            k=index+1
        k=0
        while s.find('10',k)!=-1:
            index=s.find('10',k)
            count=count+1
            i=index-1
            j=index+2
            while j<len(s) and s[i]=='1' and s[j]=='0':
                count=count+1
                i=i-1
                j=j+1
            k=index+1
        return count
